### ISSUE TITLE
ANSI Escape sequences appear in ex mode bang output.

### ISSUE THREAD

**Issue by euclio** [neovim/neovim#2765]
This is a regression from Vim.

OS: Arch Linux
Version: NVIM 0.0.0-alpha+201505292057

Steps to reproduce:
`:!echo -e "\e[0;32mV"`

Expected result:
A green 'V'

Actual result:
White " [0;32mV"

I imagine this is a result of the refactoring for the terminal UI? It makes output from CMake, etc. difficult to read.

**Comment by fwalch** [comment]
I get the same output with vim and nvim:

```
\[0;32mV
Press ENTER or type command to continue
```

**Comment by brcolow** [comment]
Mac OS X (Yosemite), VIM - Vi IMproved 7.4.712, NVIM 0.0.0-alpha+201506011617, iTerm2. I get the "[0;32mV on BOTH vim and neovim.

**Comment by euclio** [comment]
Ah, of course I had a typo in my steps to reproduce >_<. Fixed.

Also, here's a Python script taken from StackOverflow that can demonstrate the problem more readily:

```python
#!/usr/bin/env python
import sys
terse = "-t" in sys.argv[1:] or "--terse" in sys.argv[1:]
write = sys.stdout.write
for i in range(2 if terse else 10):
    for j in range(30, 38):
        for k in range(40, 48):
            if terse:
                write("\33[%d;%d;%dm%d;%d;%d\33[m " % (i, j, k, i, j, k))
            else:
                write("%d;%d;%d: \33[%d;%d;%dm Hello, World! \33[m \n" %
                      (i, j, k, i, j, k,))
        write("\n")
```

Save as mark as executable and then run in vim/nvim to see the regression.

**Comment by jbaum98** [comment]
+1 this is very important for running rspec tests from neovim because the green/red visual cues for pass/fail

**Comment by justinmk** [comment]
Why not use `:te` instead of `:!` ?

**Comment by jbaum98** [comment]
What is `:te`? My `nvim` says `The command is not available in this version.`

**Comment by justinmk** [comment]
@jbaum98 You must have a fairly old build. `:te` is the abbreviation of `:terminal`.

**Comment by jbaum98** [comment]
Thanks. I'm working on a way to use `:terminal` such that it will be as simple to run tests as with `!`, but in general `!` is much simpler and easy to use, and integrate with other tools.

**Comment by justinmk** [comment]
> in general ! is much simpler and easy to use

We want to make `:te` as easy to use as `:!`. What is easier and simpler about it?

**Comment by tarruda** [comment]
This is not a bug, nvim connects to bang commands via pipes and does not interpret escape. As suggested by @justinmk, you can use `:terminal` to achieve something similar, and with mappings you can make it just as easy as running a bang command.

**Comment by jbaum98** [comment]
@justinmk I said `:!` is simpler because it automatically displays the output on the bottom of the screen, using the minimal height, and can be dismissed with enter. I'm trying to replicate this with `:te` using splits, but `:!` has my desired functionality with no work and better than I could get it with splits, which is why I said simpler.

**Comment by justinmk** [comment]
@jbaum98 Thanks, that will be useful as we polish the user experience. Just wanted to get details on your workflow.

**Comment by euclio** [comment]
@justinmk I like that idea. I think many users expect colorized output from `:!`.

**Comment by tarruda** [comment]
> Half-baked thought: what if :! only used pipes for :r! and :w! cases, and used a mildly enhanced :te for the interactive (:!) case?

This shouldn't be hard, we basically need to decouple the terminal module to work without buffer instances(drawing directly to the screen).

> @justinmk I like that idea. I think many users expect colorized output from :!.

FWIW, gvim has the same behavior(escape sequences are printed).

**Comment by jbaum98** [comment]
Almost all vim implementations print escape sequences. 

The one exception is MacVim, which somehow delegates external commands to a shell in a different way so that we see the colors.

**Comment by justinmk** [comment]
> Almost all vim implementations print escape sequences

gvim on windows doesn't, unless you count showing a `cmd.exe` Window. What does gvim on linux do?

**Comment by akarl818** [comment]
I have the same issue.

If i do `:! git status` in vim 7.4 I get all colors correct.

But in neovim `82296510879d6e18036ba45447d348bba41be52e` I get `[31mmodified:   vimrc [m` etc.

**Comment by gonzaloserrano** [comment]
Hey guys, i had the same problem, switched to `:term` but it closes the tab you are currently editing as linked in https://github.com/neovim/neovim/issues/3276, does that happen to you too?

**Comment by akarl818** [comment]
@gonzaloserrano yes, it will use the same window as you are currently in. If you want the terminal in a new split you need to do something like `:vsp term://top`. See `:h nvim-terminal-emulator-spawning`

**Comment by ggVGc** [comment]
I think the current behavior of :! is very much breaking compatibility with vim, in unacceptable ways. Lack of color support is not great, but the main issue is that :! is running a _blocking_ command, while :terminal opens a new buffer with a terminal and the code that launched it keeps running.
Many plugins(and my vimrc) relies on :! for running an external command then then processing the output, and that currently breaks in neovim.

**Comment by mgedmin** [comment]
Note: gvim sort of does the same think as nvim here, but gvim also *tells* the running applications that it doesn't support ANSI colors (by setting TERM=dumb).  Many command-line apps (such as `git status`) check the $TERM setting and only use ANSI escapes when they're supported.

I suggest nvim set TERM=dumb as a short-term solution.

**Comment by fmoralesc** [comment]
@ggVGc See the discussion on https://github.com/neovim/neovim/issues/1496 about the behaviour of interactive programs spawned with `:!`.

**Comment by itmm** [comment]
It would be great to have color support in nvim. Not only for :! but also for :mak and similar commands. Using :terminal is more complicate to use instead.

**Comment by itmm** [comment]
Or at least set the TERM environment of the called program to signal that no escape sequences will be interpreted.

**Comment by justinmk** [comment]
closed by https://github.com/neovim/neovim/pull/7844

**Comment by eudoxos** [comment]
I am sorry to re-open this but I am not able to get colored output with the test echo (I get the same with `nvim --clean` and also with spacevim on the image):
![image](https://user-images.githubusercontent.com/1029876/71486589-63168300-2817-11ea-8f2f-50397408d4ea.png)

I also tested with `git status` which is monochrome. Is there some option I need to set to make the magic happen?

My version is:
```
VIM v0.5.0-dev
Build type: RelWithDebInfo
Lua 5.1
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-ZYxjFg/neovim-0.5.0+ubuntu1+git201912260029-5f1aec5-00e710e=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/build/neovim-ZYxjFg/neovim-0.5.0+ubuntu1+git201912260029-5f1aec5-00e710e/build/config -I/build/neovim-ZYxjFg/neovim-0.5.0+ubuntu1+git201912260029-5f1aec5-00e710e/src -I/build/neovim-ZYxjFg/neovim-0.5.0+ubuntu1+git201912260029-5f1aec5-00e710e/.deps/usr/include -I/usr/include -I/build/neovim-ZYxjFg/neovim-0.5.0+ubuntu1+git201912260029-5f1aec5-00e710e/build/src/nvim/auto -I/build/neovim-ZYxjFg/neovim-0.5.0+ubuntu1+git201912260029-5f1aec5-00e710e/build/include
Compiled by buildd@lgw01-amd64-043

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```
Do I need to set some option for this to be working?

**Comment by frivoal** [comment]
Failing for me as well, please reopen.

```
NVIM v0.4.3
Build type: Release
Lua 5.3
Compilation: /usr/bin/clang -pipe -Os -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DNVIM_UNIBI_HAS_VAR_FROM -I/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_editors_neovim/neovim/work/build/config -I/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_editors_neovim/neovim/work/neovim-0.4.3/src -I/opt/local/include -I/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_editors_neovim/neovim/work/build/src/nvim/auto -I/opt/local/var/macports/build/_opt_bblocal_var_buildworker_ports_build_ports_editors_neovim/neovim/work/build/include
Compiled by macports@mojave.internal.macports.net

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/opt/local/share/nvim"

Run :checkhealth for more info
```

### Actual results

>:!echo -e "\e[0;32mV"
> ^[[0;32mV
>
> Press ENTER or type command to continue


### Expected Results

The second line above should show a green V and no escape codes.

**Comment by bfredl** [comment]
This is a duplicate of #1496 .

**Comment by frivoal** [comment]
I don't think this is a duplicate. Maybe some root causes are shared (or maybe not), but the problem raised in this issue and the one raised in #1496 are different.

#1496 is about `:!` commands being interactive or not. I don't particularly care either way about that. I believe that issue seems to be headed towards WONTFIX, on account that using pipes, which are non interactive, is a deliberate decision. I'm OK with that.

This issue is about interpreting (or not) the ANSI escapes in the output of `:!` commands, so as to display colored text. This could still be done with pipes: we'd need something to interpret the ANSI escapes in the output of the pipe in order to display the colored text.

**Comment by bfredl** [comment]
> I don't think this is a duplicate. Maybe some root causes are shared (or maybe not), but the problem raised in this issue and the one raised in #1496 are different.

This is not a matter of opinion. The point of #1496 is to make `:!` behave like a fully featured terminal, which includes ANSI sequences. That the OP happened to use the word "interactive" to summarize this request, does not mean we need to have duplicate issues for each aspect of fully featured terminal.

>  issue seems to be headed towards WONTFIX, on account that using pipes, which are non interactive, is a deliberate decision. I'm OK with that.

Exactly where is it stated that it is WONTFIX/deliberate? You could just as well say that ANSI sequences are WONTFIX, if we are just making things up. The limiting factor here (as @justinmk clearly said in the issue) is manpower, not that we somehow actively desire to keep `:!` crippled.

> This issue is about interpreting (or not) the ANSI escapes in the output of :! commands, so as to display colored text. 

Which would be done most straightforwardly using libvterm, at which point we should use an interactive PTY as well (like gvim does). Of course a PR for a partial solution would be better than nothing, but you seem to believe that we somehow _prefer_ the partial solution to the full solution, which is simply not the case.

**Comment by frivoal** [comment]
> Exactly where is it stated that it is WONTFIX/deliberate?

That's how I read this comment: https://github.com/neovim/neovim/issues/1496#issuecomment-63691483

> This is not a bug, it is the new behavior of bang commands: We no longer spawn the program with it's stdout connected to Nvim tty, instead we open a pipe, read output and display to the user. This is the only way the bang commands will be consistent across UIs, so programs designed to be used interactively from the terminal will no longer work from inside nvim.

This does not explicitely state WONTFIX, but it describes the current behavior as intentional, so that's how I took it. I had read most, but not all, other comments in the thread (as github hid some of them due to being too many), and failed to notice it was later stated that there is a plan to fix it in 0.5.0, in a way that addresses both interactivity and ANSI escapes (as you say, by making it a full fledged terminal). Great news!

**Comment by bukzor** [comment]
> This issue is about interpreting (or not) the ANSI escapes in the output of :! commands, so as to display colored text. This could still be done with pipes: we'd need something to interpret the ANSI escapes in the output of the pipe in order to display the colored text.

I see this expressed several times during this discussion, but it's wrongheaded; the request here is to interpret the output of commands _less_, not more. Currently nvim is actively finding and pretty-printing "non-printable" characters. If instead it didn't modify the string at all before printing it, things would work as expected by OP. As proof, here's a workaround:

```vim

:let x=system('echo -e ''\033]52;c;''; base64; echo -e ''\007''', "It works!")
:echo x
:lua function rawecho(x) io.stderr:write(x) end
:lua rawecho(x)
```

The first echo shows `^[]52;c;SXQgd29ya3Mh^G` with `^[` and `^G` in a bold grey font (to indicate they've been escaped, presumably). The "raw" echo places `It works!` into the terminal's copy buffer (for most terminals).

If it's too late to amend nvim echo to pretty-print less, this should at least be noted in vim_diff.txt.

---
P.S.

A viml version of "raw echo", which also functions as a workaround for #2765:

```
  " Echo a string to the terminal without munging the escape sequences.
  function! g:Rawecho (str)
    call writefile([a:str], '/dev/fd/2', 'b')
  endfunction
```
